internal/handlers/common/distinct_iterator.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"errors"
+	"sync/atomic"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// DistinctIterator returns an iterator that returns a single document containing
+// the distinct values from documents on the specified key.
+// It will be added to the given closer.
+//
+// Next method return the distinct value of all documents in iterator.
+//
+// Close method closes the underlying iterator.
+func DistinctIterator(iter types.DocumentsIterator, closer *iterator.MultiCloser, key string) (types.DocumentsIterator, error) {
+	res := &distinctIterator{
+		iter: iter,
+	}
+
+	path, err := types.NewPathFromString(key)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	res.path = path
+
+	closer.Add(res)
+
+	return res, nil
+}
+
+// distinctIterator is returned by DistinctIterator.
+type distinctIterator struct {
+	iter types.DocumentsIterator
+	path types.Path
+	done atomic.Bool
+}
+
+// Next implements Iterator Interface.
+// The first call returns the array of distinct documents from underlyin iterator,
+// subsequent calls return ErrIteratorDone.
+// If iterator contains no document, it returns empty array.
+func (iter *distinctIterator) Next() (struct{}, *types.Document, error) {
+	var unused struct{}
+
+	distinct := types.MakeArray(0)
```

Comment:
I think the `types.Array` is not the best structure for representing a set. Is it because of the custom logic in `types.Array.Contains`? I'm not sure

Reply:
Good to know, because I had the same experience with the map key in 
 $group` https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/common/aggregations/stages/group.go#L245 :see_no_evil: